x/wasm/IBC.md

```diff
@@ -178,96 +178,98 @@ type IBCCloseChannel struct {
-// if there any point in returning a response, what does it mean?
-func IBCPacketAck(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, env IBCPacketInfo, 
-    originalMsg []byte, result []byte) error {}
-
-// same question as for IBCPacketAck
-func IBCPacketDropped(ctx sdk.Context, k *wasm.Keeper, contractAddress sdk.AccAddress, env IBCPacketInfo, 
-    originalMsg []byte, errorMsg string) error {}
-
-// do we need/want all this info?
-type IBCPacketInfo struct {
-	// local id for the Channel packet was sent/received on
-	ChannelID string
 
-    // sequence for the packet (will already be enforced in order if ORDERED channel)
-	Sequence uint64
-	
-    // Note: Timeout if guaranteed ONLY to be exceeded iff we are processing IBCPacketDropped
-    // otherwise, this is just interesting metadata
+// IBCContractCallbacks defines the methods for go-cosmwasm to interact with the wasm contract.
+// A mock contract would implement the interface to fully simulate a wasm contract's behaviour.
+type IBCContractCallbacks interface {
+  // Package livecycle
+
+  // OnIBCPacketReceive handles an incoming IBC package
+  OnIBCPacketReceive(hash []byte, params cosmwasm.Env, msg []byte, store prefix.Store, api wasm.GoAPI, querier QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm.IBCPacketReceiveResponse, uint64, error)
+  // OnIBCPacketAcknowledgement handles a IBC package execution on the counterparty chain
+  OnIBCPacketAcknowledgement(hash []byte, params cosmwasm.Env, originalData []byte, acknowledgement []byte, store prefix.Store, api wasm.GoAPI, querier QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm.IBCPacketAcknowledgementResponse, uint64, error)
+  // OnIBCPacketTimeout reverts state when the IBC package execution does not come in time
+  OnIBCPacketTimeout(hash []byte, params cosmwasm.Env, msg []byte, store prefix.Store, api wasm.GoAPI, querier QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm.IBCPacketTimeoutResponse, uint64, error)
+  // channel livecycle
+
+  // OnIBCChannelOpen does the protocol version negotiation during channel handshake phase
+  OnIBCChannelOpen(hash []byte, params cosmwasm.Env, order channeltypes.Order, version string, store prefix.Store, api wasm.GoAPI, querier QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm.IBCChannelOpenResponse, uint64, error)
+  // OnIBCChannelConnect callback when a IBC channel is established
+  OnIBCChannelConnect(hash []byte, params cosmwasm.Env, counterpartyPortID, counterpartyChannelID string, store prefix.Store, api wasm.GoAPI, querier QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm.IBCChannelConnectResponse, uint64, error)
+  // OnIBCChannelConnect callback when a IBC channel is closed
+  OnIBCChannelClose(ctx sdk.Context, hash []byte, params cosmwasm.Env, counterpartyPortID, counterpartyChannelID string, store prefix.Store, api wasm.GoAPI, querier QueryHandler, meter sdk.GasMeter, gas uint64) (*cosmwasm.IBCChannelCloseResponse, uint64, error)
```

Comment:
`counterpartyPortID, counterpartyChannelID string` do we need these? We get our own side in Env and can query counterParty if needed. We should just use the channelID locally to look up everything.

I can see passing this explicitly in `OnIBCChannelConnect` to provide all relevant info on connect to be use for setup if needed. But once it is set up, do we usually care? And if so we can query